### PR TITLE
misc: Pass SSAValue or IRNode for structural equality checking

### PR DESCRIPTION
**Pull request by superlopuh**
This was actually kind of assumed by the code checking for structural equivalence. The parent is currently `None`, which feels about right. We could probably make it the same as `owner`, although I'm not sure it's worth doing if everything works.

### CODE REVIEW

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -45,8 +45,7 @@ def verify(self, attr: Attribute) -> None:
 class ArrayAttr(GenericData[tuple[AttributeCovT, ...]]):
     name: str = "array"
 
-    def __init__(self: ArrayAttr[AttributeCovT],
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:48]: not sure why the self was annotated here, `Pylance` suddenly started screaming after the change in ir.py, and this change calmed it down...

**Review by webmiche — COMMENTED** [review]
So, I am against this change. IMO, the term `IRNode` has been ill-advised from the start, I think of it more as an `IRConstruct`. Sure, you can abstract objects in a data structure as a graph, but I am not sure what that actually gives you when doing compilation.

IMO, there is two kinds of important graphs in an SSA IR. CFGs (i.e. amade up of all of the structures we currrently refer to as IRNodes) and def-use chains. IMO, this change mixes the two and I am not sure what we gain from that. Maybe I am missing something, but this is only introducing several methods on SSAValue, that does not really have a use (at least now) and to me should not be there.

If I am missing something, I would be very interested to learn that. However, atm, I give this change a weak 👎

**Comment by superlopuh** [comment]
To be honest, I'm with you here. I'd love feedback on a better way to make this change then. One possible way to do it would be to have an object wrapping the `dict[IRNode,IRNode]` in the structural equality method, that would have an API for unification. I'll try that. On the other hand, I don't know if it's worth doing this for this PR, vs a follow-up PR.

**Comment by webmiche** [comment]
Ah I think I see your problem. The type annotations seem to be wrong, the context maps IRNodes to IRNodes, but is then populated with SSAValues. Can we make this just a union of IRNode and SSAValue?

File: xdsl/ir.py [hunk 2]
```diff
@@ -431,12 +431,19 @@ def get_toplevel_object(self) -> IRNode:
...
-            other: IRNode,
-            context: Optional[dict[IRNode, IRNode]] = None) -> bool:
+        self,
+        other: IRNode,
+        context: dict[IRNode | SSAValue, IRNode | SSAValue] | None = None
+    ) -> bool:
         """Check if two IR nodes are structurally equivalent."""
         ...
 
+    def __eq__(self, other: object) -> bool:
```
Comment (math-fehr) [inline xdsl/ir.py:441]: Why is this required?
Should we mark them as `abstractmethod` instead?

**Review by math-fehr — APPROVED** [review]
Yes, I also disagree on making `SSAValue` an `IRNode`.
And I agree with this change ;)

**Review by webmiche — APPROVED** [review]
Yes, this looks better to me. Maybe @martin-luecke has a different opinion though (that is his code after all).

Modulo Mathieus comment, this LGTM now.

**Review by martin-luecke — APPROVED** [review]
@webmiche I guess as soon as its merged it is everyone's code. :wink:  
This looks good to me!
The only question is: Is it maybe better design to have a `nodeContext: dict[IRNode, IRNode]` and a `valueContext: dict[SSAValue, SSAValue]`?
But I am fine with this

**Comment by webmiche** [comment]
Yep, I was thinking about that too. Not sure what we gain from that though, so I am fine with leaving as is.

**Comment by superlopuh** [comment]
I'd like to revisit this in the future, feels like moving some more logic around might also be a good idea. The aim of this PR was to minimise code changes while making the logic more explicit through types, so merged as-is.
